internal/services/securitycenter/security_center_storage_defender_resource.go

```diff
@@ -0,0 +1,314 @@
+package securitycenter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-azure-helpers/lang/pointer"
+	"github.com/hashicorp/go-azure-helpers/lang/response"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonids"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/security/2022-12-01-preview/defenderforstorage"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-azurerm/helpers/tf"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+)
+
+type StorageDefenderResource struct{}
+
+type StorageDefenderModel struct {
+	StorageAccountId                 string `tfschema:"storage_account_id"`
+	Enabled                          bool   `tfschema:"enabled"`
+	OverrideSubscriptionSettings     bool   `tfschema:"override_subscription_settings_enabled"`
+	MalwareScanningOnUploadEnabled   bool   `tfschema:"malware_scanning_on_upload_enabled"`
+	MalwareScanningOnUploadCapPerMon int64  `tfschema:"malware_scanning_on_upload_cap_gb_per_month"`
+	SensitiveDataDiscoveryEnabled    bool   `tfschema:"sensitive_data_discovery_enabled"`
+}
+
+var _ sdk.ResourceWithUpdate = StorageDefenderResource{}
+
+func (s StorageDefenderResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return commonids.ValidateScopeID
+}
+
+func (s StorageDefenderResource) ModelObject() interface{} {
+	return &StorageDefenderModel{}
+}
+
+func (s StorageDefenderResource) ResourceType() string {
+	return "azurerm_security_center_storage_defender"
+}
+
+func (s StorageDefenderResource) Arguments() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"storage_account_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: commonids.ValidateStorageAccountID,
+		},
+
+		"enabled": {
+			Type:     pluginsdk.TypeBool,
+			Required: true,
+		},
+
+		"override_subscription_settings_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+
+		"malware_scanning_on_upload_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+
+		"malware_scanning_on_upload_cap_gb_per_month": {
+			Type:     pluginsdk.TypeInt,
+			Optional: true,
+			Default:  -1,
+			ValidateFunc: func(i interface{}, s string) (warnings []string, errors []error) {
+				// it requires -1 or greater than 0
+				v, ok := i.(int)
+				if !ok {
+					errors = append(errors, fmt.Errorf("expected type of %s to be integer", s))
+					return warnings, errors
+				}
+
+				if v == -1 {
+					return warnings, errors
+				}
+
+				return validation.IntAtLeast(-1)(i, s)
+			},
+		},
```

Comment:
What does `-1` mean in this case? Also what would `0` mean?

Reply:
the service does not accept 0, and -1 means no limit on that cap. Do we need to map user input `0` to payload `-1`?

Reply:
I think this can be more clearly/simply achieved with:
```suggestion
		"malware_scanning_on_upload_cap_gb_per_month": {
			Type:     pluginsdk.TypeInt,
			Optional: true,
			Default:  -1,
			ValidateFunc: validation.Any(
									validation.IntAtLeast(1),
									validation.IntInSlice([]int{-1}),
								),
		},
```